Public/Src/Cache/ContentStore/Vfs/VirtualizedContentSession.cs

```diff
@@ -24,89 +26,127 @@
         {
-            return InnerSession.OpenStreamAsync(operationContext, contentHash, operationContext.Token, urgencyHint);
+            return _innerSession.OpenStreamAsync(operationContext, contentHash, operationContext.Token, urgencyHint);
         }
 
         /// <inheritdoc />
         protected override Task<PinResult> PinCoreAsync(OperationContext operationContext, ContentHash contentHash, UrgencyHint urgencyHint, Counter retryCounter)
         {
-            return InnerSession.PinAsync(operationContext, contentHash, operationContext.Token, urgencyHint);
+            return _innerSession.PinAsync(operationContext, contentHash, operationContext.Token, urgencyHint);
         }
 
         /// <inheritdoc />
         protected override Task<IEnumerable<Task<Indexed<PinResult>>>> PinCoreAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint, Counter retryCounter, Counter fileCounter)
         {
-            return InnerSession.PinAsync(operationContext, contentHashes, operationContext.Token, urgencyHint);
+            return _innerSession.PinAsync(operationContext, contentHashes, operationContext.Token, urgencyHint);
         }
 
         /// <inheritdoc />
-        protected override Task<PlaceFileResult> PlaceFileCoreAsync(OperationContext operationContext, ContentHash contentHash, AbsolutePath path, FileAccessMode accessMode, FileReplacementMode replacementMode, FileRealizationMode realizationMode, UrgencyHint urgencyHint, Counter retryCounter)
+        protected async override Task<PlaceFileResult> PlaceFileCoreAsync(OperationContext operationContext, ContentHash contentHash, AbsolutePath path, FileAccessMode accessMode, FileReplacementMode replacementMode, FileRealizationMode realizationMode, UrgencyHint urgencyHint, Counter retryCounter)
+        {
+            if (replacementMode != FileReplacementMode.ReplaceExisting && _fileSystem.FileExists(path))
+            {
+                if (replacementMode == FileReplacementMode.SkipIfExists)
+                {
+                    return new PlaceFileResult(PlaceFileResult.ResultCode.NotPlacedAlreadyExists);
+                }
+                else if (replacementMode == FileReplacementMode.FailIfExists)
+                {
+                    return new PlaceFileResult(
+                        PlaceFileResult.ResultCode.Error,
+                        $"File exists at destination {path} with FailIfExists specified");
+                }
+            }
+
+            var virtualPath = _contentManager.ToVirtualPath(path);
+            if (virtualPath == null)
+            {
+                return await _innerSession.PlaceFileAsync(operationContext, contentHash, path, accessMode, replacementMode, realizationMode, operationContext.Token, urgencyHint);
+            }
+
+            _contentManager.Tree.AddFileNode(virtualPath, new VfsFilePlacementData(contentHash, realizationMode, accessMode));

```

Comment:
>AddFileNode [](start = 33, length = 11)

So is this just placing the content in memory?

Reply:
Correct.

---
In reply to: [298369962](https://github.com/microsoft/BuildXL/pull/519#discussion_r298369962) [](ancestors = 298369962)